compiler/rustc_const_eval/src/const_eval/valtrees.rs

```diff
@@ -0,0 +1,479 @@
+
+            let const_val = op_to_const(&ecx, &place.into());
+            debug!(?const_val);
+
+            const_val
+        }
+        ty::Never
+        | ty::FnDef(..)
+        | ty::Error(_)
+        | ty::Foreign(..)
+        | ty::Infer(ty::FreshIntTy(_))
+        | ty::Infer(ty::FreshFloatTy(_))
+        | ty::Projection(..)
+        | ty::Param(_)
+        | ty::Bound(..)
+        | ty::Placeholder(..)
+        | ty::Opaque(..)
+        | ty::Infer(_)
+        | ty::Closure(..)
+        | ty::Generator(..)
+        | ty::GeneratorWitness(..)
+        | ty::FnPtr(_)
+        | ty::RawPtr(_)
+        | ty::Str
+        | ty::Slice(_)
+        | ty::Dynamic(..) => bug!("no ValTree should have been created for type {:?}", ty.kind()),
+    }
+}
+
+// FIXME Needs a better/correct name
+#[instrument(skip(ecx), level = "debug")]
+fn fill_place_recursively<'tcx>(
+    ecx: &mut CompileTimeEvalContext<'tcx, 'tcx>,
+    place: &mut MPlaceTy<'tcx>,
+    valtree: ty::ValTree<'tcx>,
+    ty: Ty<'tcx>,
+) {
+    // This will match on valtree and write the value(s) corresponding to the ValTree
+    // inside the place recursively.
+
+    let tcx = ecx.tcx.tcx;
+
+    match ty.kind() {
+        ty::Bool | ty::Int(_) | ty::Uint(_) | ty::Float(_) | ty::Char => {
+            let scalar_int = valtree.unwrap_leaf();
+            debug!("writing trivial valtree {:?} to place {:?}", scalar_int, place);
+            ecx.write_immediate(
+                Immediate::Scalar(ScalarMaybeUninit::Scalar(scalar_int.into())),
+                &(*place).into(),
+            )
+            .unwrap();
+        }
+        ty::Ref(_, inner_ty, _) => {
+            match inner_ty.kind() {
+                ty::Slice(_) | ty::Str => {
+                    let slice_ty = match inner_ty.kind() {
+                        ty::Slice(slice_ty) => *slice_ty,
+                        ty::Str => tcx.mk_ty(ty::Uint(ty::UintTy::U8)),
+                        _ => bug!("expected ty::Slice | ty::Str"),
+                    };
+                    debug!(?slice_ty);
+
+                    let valtrees = valtree.unwrap_branch();
+                    debug!(?valtrees);
+                    let len = valtrees.len();
+                    debug!(?len);
+
+                    // create a place for the underlying array
+                    let arr_ty = tcx.mk_array(slice_ty, len as u64);
+                    let mut arr_place =
+                        create_mplace_from_layout(ecx, ty::ParamEnv::empty().and(arr_ty));
+                    debug!(?arr_place);
+
+                    // Insert elements of `arr_valtree` into `place`
+                    fill_place_recursively(ecx, &mut arr_place, valtree, arr_ty);
+                    dump_place(&ecx, arr_place.into());
+
+                    // Now we need to create a `ScalarPair` from the filled `place`
+                    // and write that into `place`
+                    let (alloc_id, offset) = arr_place.mplace.ptr.into_parts();
```

Comment:
Instead of taking the pointer apart and putting it back together, just use the pointer directly.

Also use the deref impl to get the field instead of accessing the field

Reply:
The destructuring was unfortunately necessary since the pointer corresponding to the place has type `Pointer<Option<AllocId>>`, but `ScalarMaybeUninit::from_pointer` requires `Pointer<AllocId>>` (as does the deref call). Didn't find any way to unwrap the inner alloc_id.

Reply:
This was also why I had to make `offset` of `Pointer` non-private (which is supposed to be private). There has to be a way to do this, since I would assume `Pointer<Option<AllocId>>` is encountered quite often.

Reply:
There is `ScalarMaybeUninit::from_maybe_pointer` for this exact case. :)